File: typescript/lazyk.ts
export enum EType {
  App = 0,
  S,
  S1,
  S2,
  K,
  K1,
  I,
  I1,
  Iota,
  False,
  CNum,
  CNum1,
  Inc,
  Num,
  Read,
}

export class Expr {
  type: number;
  intval: number;
  arg1?: Expr;
  arg2?: Expr;
  constructor(t: EType, a1?: Expr, a2?: Expr, intval: number = 0) {
    this.type = t;
    this.arg1 = a1;
    this.arg2 = a2;
    this.intval = intval;
  }

  // static members
  static S = new Expr(EType.S);
  static K = new Expr(EType.K);
  static I = new Expr(EType.I);
  static Iota = new Expr(EType.Iota);
  static False = new Expr(EType.False);
  static Inc = new Expr(EType.Inc);
  private static _cnums = new Array(257);
  private static _succ = Expr.S.apply(Expr.S.apply(Expr.K.apply(Expr.S)).apply(Expr.K))
  static CNum(n: number): Expr {
    return Expr._cnums[n] ||= new Expr(EType.CNum, undefined, undefined, n);
  }
  static Num(n: number): Expr {
    return new Expr(EType.Num, undefined, undefined, n);
  }
  static Num0 = Expr.Num(0);

  // methods
  apply(a: Expr): Expr {
    return new Expr(EType.App, this, a);
  }

  toString(): string {
    switch (this.type) {
      case EType.S1, EType.K1, EType.I1:
        return `(${EType[this.type][0]} ${this.arg1?.toString()})`;
      case EType.App:
        return `(${this.arg1?.toString()} ${this.arg2?.toString()})`;
      case EType.S2:
        return `(S ${this.arg1?.toString()} ${this.arg2?.toString()})`;
      case EType.CNum, EType.Num:
        return EType[this.type] + `#${this.intval}`;
      case EType.CNum1:
        return `(CNum#${this.intval} ${this.arg1?.toString()})`;
      default:
        return EType[this.type];
    }
  }

  dropI1(): Expr {
    let cur: Expr = this;
    while (cur.type === EType.I1) {
      cur = cur.arg1 as Expr;
    }
    return cur;
  }
}

export class Parser {
  private _src: string;
  private _p: number;
  private _unget: string[];
  constructor(src: string) {
    this._src = src;
    this._p = 0;
    this._unget = [];
  }

  private nextc(): string {
    if (this._unget.length > 0) return this._unget.pop() as string;
    let c: string;
    do {
      if (this._p === this._src.length) return "";
      c = this._src[this._p++];
      if (c === "#") {
        do {
          if (this._p === this._src.length) return "";
          c = this._src[this._p++];
        } while (c !== "\n");
      }
    } while (c.match(/\s/));
    return c;
  }

  private ungetc(c: string): void {
    this._unget.push(c);
  }

  readCC(closingchar: string): Expr {
    let c = this.nextc();
    if (c === closingchar) return Expr.I;
    this.ungetc(c);
    let e: Expr = this.readOne(false);
    while ((c = this.nextc()) !== closingchar) {
      this.ungetc(c);
      e = e.apply(this.readOne(false));
    }
    return e;
  }

  readOne(i_is_iota: boolean): Expr {
    const c = this.nextc();
    switch (c) {
      case "`":
      case "*":
        return this.readApp(c);
      case "(":
        return this.readCC(")");
      case "S":
      case "s":
        return Expr.S;
      case "K":
      case "k":
        return Expr.K;
      case "I":
        return Expr.I;
      case "i":
        return (i_is_iota ? Expr.Iota : Expr.I);
      case "0":
      case "1":
        return this.readJot(c);
      case "":
        throw new Error("unexpected EOF");
      default:
        throw new Error(`unexpected char: ${c}`);
    }
  }

  readApp(c0: string): Expr {
    const stack: { in_iota: boolean; buf: Expr[] }[] = [];
    let buf: Expr[] = [];
    let in_iota = c0 == "*";
    while (true) {
      const c = this.nextc();
      if (c === "`" || c === "*")
      {
        stack.push({ in_iota, buf });
        in_iota = c === "*";
        buf = [];
      } else {
        this.ungetc(c);
        buf.push(this.readOne(in_iota));
      }
      while (buf.length === 2) {
        const e = buf[0].apply(buf[1]);
        if (stack.length === 0) return e;
        const s = stack.pop();
        if (s) {
          in_iota = s.in_iota;
          buf = s.buf;
          buf.push(e);
        }
      }
    }
  }

  readJot(c0: string): Expr {
    let e = Expr.I;
    let c = c0;
    while (c === "0" || c === "1") {
      if (c === "0") {
        e = e.apply(Expr.S).apply(Expr.K);
      } else {
        e = Expr.S.apply(Expr.K.apply(e));
      }
      c = this.nextc();
    }
    this.ungetc(c);
    return e;
  }

  static parse(src: string): Expr {
    return new Parser(src).readCC("");
  }
}

export interface IObase {
  readsync(buf: Uint8Array): number | null;
  writesync(buf: Uint8Array): void;
}

export class LazyK {
  private readonly _io: IObase;
  private _inbuf: number[];
  private _iniseof: boolean;
  private _outbuf: number[];

  constructor(io: IObase) {
    this._io = io;
    this._inbuf = [];
    this._iniseof = false;
    this._outbuf = [];
  }

  private read = (): number => {
    if (!this._iniseof && this._inbuf.length === 0) {
      this.flush();
      const buf = new Uint8Array(1024);
      const len = this._io.readsync(buf);
      if (len === null) {
        this._iniseof = true;
      } else {
        this._inbuf = Array.from(buf.slice(0, len));
      }
    }
    if (this._iniseof) {
      return 256;
    } else {
      return this._inbuf.shift() as number;
    }
  };

  private flush(): void {
    const a = new Uint8Array(this._outbuf);
    this._outbuf = [];
    this._io.writesync(a);
  }

  private write(c: number): void {
    this._outbuf.push(c);
    if (c === 10) this.flush();
  }

  churchNumToInt(cn: Expr): number {
    const n = this.eval(cn.apply(Expr.Inc).apply(Expr.Num0));
    if (n.type === EType.Num) {
      return n.intval;
    } else {
      throw `invalid output format: ${this.toString()}`;
    }
  }

  eval(e: Expr): Expr {
    const stack: Expr[] = [];
    let cur: Expr = e;
    while (true) {
      cur = cur.dropI1();
      while (cur.type === EType.App) {
        stack.push(cur);
        cur = (cur.arg1 as Expr).dropI1();
      }
      if (stack.length === 0) return cur;
      const a = cur;
      cur = stack.pop() as Expr;
      cur.arg1 = a;
      this.evalPrimitive(cur);
    }
  }

  private evalPrimitive(e: Expr): void {
    let lhs: Expr = e.arg1 as Expr;
    let rhs: Expr = e.arg2 as Expr;
    let ch: Expr;
    let readnext: Expr;
    e.arg1 = e.arg2 = undefined;
    switch (lhs.type) {
      case EType.False:
        e.type = EType.I;
        break;
      case EType.I:
        e.type = EType.I1;
        e.arg1 = rhs;
        break;
      case EType.K:
        e.type = EType.K1;
        e.arg1 = rhs;
        break;
      case EType.K1:
        e.type = EType.I1;
        e.arg1 = lhs.arg1;
        break;
      case EType.S:
        e.type = EType.S1;
        e.arg1 = rhs;
        break;
      case EType.S1:
        e.type = EType.S2;
        e.arg1 = lhs.arg1;
        e.arg2 = rhs;
        break;
      case EType.S2:
        e.arg1 = (lhs.arg1 as Expr).apply(rhs);
        e.arg2 = (lhs.arg2 as Expr).apply(rhs);
        break;
      case EType.Iota:
        e.arg1 = rhs.apply(Expr.S);
        e.arg2 = Expr.K;
        break;
      case EType.Read:
        ch = Expr.CNum(this.read());
        readnext = new Expr(EType.Read);
        e.arg1 = rhs.apply(ch);
        e.arg2 = readnext;
        lhs.type = EType.S2;
        lhs.arg1 = new Expr(EType.S2, Expr.I, new Expr(EType.K1, ch));
        lhs.arg2 = new Expr(EType.K1, readnext);
        lhs.intval = 0;
        break;
      case EType.CNum:
        e.type = EType.CNum1;
        e.arg1 = rhs;
        e.intval = lhs.intval;
        break;
      case EType.CNum1:
        if ((lhs.arg1 as Expr).type === EType.Inc && rhs.type === EType.Num) {
          e.type = EType.Num;
          e.intval = lhs.intval + rhs.intval;
        } else {
          const f = lhs.arg1 as Expr;
          let x = rhs;
          for (let i = 0; i < lhs.intval; i++) {
            x = f.apply(x);
          }
          e.type = x.type;
          e.arg1 = x.arg1;
          e.arg2 = x.arg2;
          e.intval = x.intval;
        }
        break;
      case EType.Inc:
        if (rhs.type !== EType.Num) {
          rhs = this.eval(rhs);
        }
        if (rhs.type === EType.Num) {
          e.type = EType.Num;
          e.intval = rhs.intval + 1;
        } else {
          throw new Error(
            `invalid output format (attempted to apply inc to a non-number): (${rhs.toString()})`,
          );
        }
        break;
      case EType.Num:
        throw new Error(
          `invalid output format (${lhs.toString()} is not applicable)`,
        );
      default:
        throw new Error(
          `unexpected state (lhs=${lhs.toString()} rhs=${rhs.toString()})`,
        );
    }
  }

  printList(lst: Expr): number {
    const carch = (e: Expr) => this.churchNumToInt(e.apply(Expr.K));
    const cdr = (e: Expr) => this.eval(e.apply(Expr.False));
    let c: number;
    while ((c = carch(lst)) < 256) {
      this.write(c);
      lst = cdr(lst);
    }
    this.flush();
    return c - 256;
  }

  run(src: string): number {
    const parsed = Parser.parse(src);
    try {
      return this.printList(parsed.apply(new Expr(EType.Read)));
    } catch (e) {
      this.flush();
      console.log(e);
      return 1;
    }
  }
}

function main() {
  if (Deno.args.length === 0) {
    console.log("usage: deno run --allow-read lazyk.ts <source.lazy>");
    Deno.exit(1);
  }
  const src = Deno.readTextFileSync(Deno.args[0]);
  const io = {
    readsync: (buf: Uint8Array) => {
      return Deno.stdin.readSync(buf);
    },
    writesync: (buf: Uint8Array) => {
      Deno.stdout.writeSync(buf);
    },
  };
  Deno.exit(new LazyK(io).run(src));
}

if (import.meta.main) {
  main();
}

File: ocaml/lazyk.ml

type expr =
  | L of expr Lazy.t
  | F of (expr -> expr)
  | N of int

let rec eval = function
  | L(l) -> eval (Lazy.force l)
  | e -> e

let eval_app lhs rhs =
  match eval lhs with
  | F(f) -> f rhs
  | N(_) -> failwith "invalid output format (attempted to apply num)"
  | _ -> failwith "unreachable code"

let apply f a = L(lazy (eval_app f a))

let s = F(fun x -> F(fun y -> F(fun z -> apply (apply x z) (apply y z))))
let k = F(fun x -> F(fun _ -> x))
let i = F(fun x -> x)
let iota = F(fun x -> apply (apply x s) k)
let ctrue = k
let cfalse = F(fun _ -> i)
let inc = F(fun e ->
              match eval e with
              | N(n) -> N(n + 1)
              | _ -> failwith "invalid output format"
            )

let num0 = N(0)
let succ = apply s (apply (apply s (apply k s)) k)
let cnums = Seq.iterate (apply succ) cfalse |> Seq.take 257 |> Array.of_seq

let int_of_cnum cn =
  match eval (apply (apply cn inc) num0) with
  | N(n) -> n
  | _ -> failwith "invalid output format"

let rec make_reader r initial_arg =
  let f () =
    let (c, next_arg) = r initial_arg in
    let cn = cnums.(c) in
    F(fun a -> apply (apply a cn) (make_reader r next_arg)) in
  L(lazy (f ()))

let chan_io ic oc =
  let reader () =
    flush oc;
    let b = try input_byte ic with End_of_file -> 256 in
    (b, ()) in
  let writer c =
    output_byte oc c;
    if c = 10 then
      flush oc in
  (make_reader reader (), writer)

let mem_io in_bytes out_buffer =
  let ilen = Bytes.length in_bytes in
  let reader i =
    if i = ilen then (256, i)
    else (Bytes.get_uint8 in_bytes i, i + 1) in
  let writer c =
    Buffer.add_int8 out_buffer c in
  (make_reader reader 0, writer)

let print_list writer =
  let rec loop l =
    let i = int_of_cnum (apply l ctrue) in
    if i >= 256 then i - 256
    else (writer i; loop (apply l cfalse)) in
  loop


let parse app s k i iota src =
  let rec skip_line = function
    | [] -> []
    | '\n' :: t -> t
    | _ :: t -> skip_line t in
  let isspace c =
    let i = Char.code c in
    (i >= 9 && i <= 13) || i = 32 in
  let rec skip_ws = function
    | [] -> []
    | (h :: t) as l ->
       if isspace h then skip_ws t
       else if h = '#' then skip_line t |> skip_ws
       else l in
  let rec read_cc l closingchar =
    let rec loop l acc =
      match skip_ws l, closingchar with
      | [], None -> (acc, [])
      | [], Some(_) -> failwith "unexpected EOF"
      | h :: t, Some(c) when h = c -> (acc, t)
      | l, _ ->
         let (e, r) = read_one l false in
         loop r (if acc == i then e else app acc e) in
    loop l i
  and read_one l i_is_iota =
    match skip_ws l with
    | [] -> failwith "unexpected EOF"
    | '`' :: t ->
       let (f, t) = read_one t false in
       let (a, t) = read_one t false in
       (app f a, t)
    | '*' :: t ->
       let (f, t) = read_one t true in
       let (a, t) = read_one t true in
       (app f a, t)
    | '(' :: r -> read_cc r (Some(')'))
    | (h :: t) as l -> (
      match h with
      | 'S' | 's' -> (s, t)
      | 'K' | 'k' -> (k, t)
      | 'I' -> (i, t)
      | 'i' -> ((if i_is_iota then iota else i), t)
      | '0' | '1' -> read_jot l
      | u -> failwith (Printf.sprintf "unexpected token: '%c'" u)
    )
  and read_jot l =
    let rec loop l e =
      match skip_ws l with
      | '0' :: t -> loop t (app (app e s) k)
      | '1' :: t -> loop t (app s (app k e))
      | r -> (e, r) in
    loop l i in
  let (e, _) = read_cc (String.to_seq src |> List.of_seq) None in
  e


let run_lazyk srcstr r w =
  try
    let prog = parse apply s k i iota srcstr in
    let reuslt = print_list w (apply prog r) in
    Ok(reuslt)
  with e ->
    Error(e)

let run_lazyk_chan srcstr ic oc =
  let (r, w) = chan_io ic oc in
  let result = run_lazyk srcstr r w in
  flush oc;
  result

(* for test *)
let run_lazyk_memio srcstr in_bytes =
  let outbuf = Buffer.create 1024 in
  let (r, w) = mem_io in_bytes outbuf in
  let result = run_lazyk srcstr r w in
  (result, Buffer.to_bytes outbuf)
